src/grt/src/GlobalRouter.cpp

```diff
@@ -924,6 +924,69 @@ void GlobalRouter::computeTrackConsumption(
   }
 }
 
+std::vector<int> GlobalRouter::findTransitionLayers()
+{
+  odb::dbTech* tech = db_->getTech();
+  int max_layer = std::max(max_routing_layer_, max_layer_for_clock_);
+  std::map<int, odb::dbTechVia*> default_vias = getDefaultVias(max_layer);
+  std::vector<int> transition_layers;
+  for (const auto [layer, via] : default_vias) {
+    odb::dbTechLayer* tech_layer = tech->findRoutingLayer(layer);
+    bool vertical = tech_layer->getDirection() == odb::dbTechLayerDir::VERTICAL;
+    int via_width = 0;
+    for (const auto box : default_vias[layer]->getBoxes()) {
+      if (box->getTechLayer()->getRoutingLevel() == layer) {
+        via_width = vertical ? box->getWidth() : box->getLength();
+        break;
+      }
+    }
+
+    int track_pitch = grid_->getTrackPitches()[layer - 1];
+    if ((static_cast<double>(via_width) / track_pitch) > 0.8) {
+      transition_layers.push_back(layer);
+    }
+  }
+
+  return transition_layers;
+}
+
+void GlobalRouter::adjustTransitionLayers(
+    const std::vector<int>& transition_layers,
+    std::map<int, std::vector<odb::Rect>>& layer_obs_map)
+{
+  odb::dbTech* tech = db_->getTech();
+  for (int layer : transition_layers) {
+    odb::dbTechLayer* tech_layer = tech->findRoutingLayer(layer);
+    for (const auto& obs : layer_obs_map[layer - 1]) {
+      odb::Rect first_tile_bds, last_tile_bds;
+      odb::Point first_tile, last_tile;
+      grid_->getBlockedTiles(
+          obs, first_tile_bds, last_tile_bds, first_tile, last_tile);
+      if (first_tile.x() != last_tile.x() || first_tile.y() != last_tile.y()) {
+        if (tech_layer->getDirection() == odb::dbTechLayerDir::HORIZONTAL) {
+          for (int y = first_tile.y(); y < last_tile.y(); y++) {
+            for (int x = first_tile.x(); x < last_tile.x(); x++) {
+              int edge_cap = fastroute_->getEdgeCapacity(x, y, x + 1, y, layer);
+              int new_cap = std::floor(static_cast<float>(edge_cap) * (0.5));
+              new_cap = edge_cap > 0 ? std::max(new_cap, 1) : new_cap;
+              fastroute_->addAdjustment(x, y, x + 1, y, layer, new_cap, true);
+            }
+          }
+        } else {
+          for (int x = first_tile.x(); x < last_tile.x(); x++) {
+            for (int y = first_tile.y(); y < last_tile.y(); y++) {
+              int edge_cap = fastroute_->getEdgeCapacity(x, y, x, y + 1, layer);
+              int new_cap = std::floor(static_cast<float>(edge_cap) * (0.5));
+              new_cap = edge_cap > 0 ? std::max(new_cap, 1) : new_cap;
+              fastroute_->addAdjustment(x, y, x, y + 1, layer, new_cap, true);
+            }
+          }
+        }
```

Comment:
I think you could combine these two conditions into one.

```c++
for (int x = first_tile.x(); x < last_tile.x(); x++) {
    for (int y = first_tile.y(); y < last_tile.y(); y++) {
        int end_x = x;
        int end_y = y;
        if (tech_layer->getDirection() == odb::dbTechLayerDir::HORIZONTAL) {
            end_x = x + 1;
        } else {
            end_y = y + 1;
        }

        int edge_cap = fastroute_->getEdgeCapacity(x, y, end_x, end_y, layer);
        int new_cap = std::floor(static_cast<float>(edge_cap) * (0.5));
        new_cap = edge_cap > 0 ? std::max(new_cap, 1) : new_cap;
        fastroute_->addAdjustment(x, y, end_x, end_y, layer, new_cap, true);
    }
}
```

Reply:
Thanks, that's much better!